### PR TITLE
Freezing Module containing fork subgraphs

### PR DESCRIPTION
**Pull request by bzinodev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38830 Preserve user specified attributes and methods
* **#37044 Freezing Module containing fork subgraphs**

This patch applies freezing to module containing fork subgraphs.

Differential Revision: [D21956334](https://our.internmc.facebook.com/intern/diff/D21956334)

### CODE REVIEW

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 1]
```diff
@@ -332,12 +395,32 @@ class AttributePropagator {
               auto attr = module.attr(name);
               insertMutableAttr(name, attr, mptr);
               if (attr.isModule()) {
-                modules.insert(attr.toModule()._ivalue());
+                auto module = attr.toModule();
```
Comment (eellison) [inline torch/csrc/jit/passes/freeze_module.cpp:398]: why is this change needed? if we insert the module and it already exists that should be fine

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 2]
```diff
@@ -332,12 +395,32 @@ class AttributePropagator {
...
+                  }
+                  modules.insert(module._ivalue());
+                }
               }
               break;
             }
           }
+        } else if (n->kind() == prim::fork) {
+          applyToSubGraph(n, graph, [this](std::shared_ptr<Graph>& subgraph) {
+            recordReferencedAttrs(subgraph);
```
Comment (eellison) [inline torch/csrc/jit/passes/freeze_module.cpp:413]: Why not just directly call `recordReferencedAttrs` here ?

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 3]
```diff
@@ -351,10 +434,14 @@ class AttributePropagator {
...
     auto type = module.type();
     size_t N = type->numAttributes();
-    if (moduleEscapes(module, graph)) {
+    /*if (moduleEscapes(module, graph)) {
       // Perserve all its attributes and methods.
       attrsToKeep_[type].insert(N);
       return;
+    }*/
+    if (attrsToKeep_[type].count(N)) {
+      // Module escapes. Preserve all its attributes and methods.
```
Comment (eellison) [inline torch/csrc/jit/passes/freeze_module.cpp:443]: What are these changes ?

**Review by eellison — COMMENTED** [review]
Changes mostly look fine but it'd be easier to review with debug code removed

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 4]
```diff
@@ -16,10 +16,33 @@ class AttributePropagator {
...
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::fork) {
+          auto subgraph = n->g(attr::Subgraph);
+          optimizeSubGraphs(subgraph, func);
```
[the remark this answers is not in the document]
Comment (bzinodev) [reply torch/csrc/jit/passes/freeze_module.cpp:33]: I ran performance on several and i do see regressions. We may need to run optimizations. I am looking into it to understand why there are regressions

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 5]
```diff
@@ -286,11 +318,42 @@ class AttributePropagator {
...
     }
   }
 
+  void applyToSubGraph(
+      Node* n,
+      std::shared_ptr<Graph>& graph,
+      std::function<void(std::shared_ptr<Graph>&)> func) {
+    TORCH_CHECK(n->kind() == prim::fork);
+    auto attrModule = module_;
+    auto node = n->inputs()[0]->node();
```
[the remark this answers is not in the document]
Comment (bzinodev) [reply torch/csrc/jit/passes/freeze_module.cpp:336]: It is a module (this in c++) but it is not enforced. 
This is how the attributes are accessed from fork subgraph

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 6]
```diff
@@ -286,11 +318,42 @@ class AttributePropagator {
...
+    // TODO: Short cut
+    TORCH_INTERNAL_ASSERT(node->kind() == prim::GetAttr);
```
[the remark this answers is not in the document]
Comment (bzinodev) [reply torch/csrc/jit/passes/freeze_module.cpp:338]: Yes, that is correct. I changed the code to if ( ...=! GetAttr) return;

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 7]
```diff
@@ -332,12 +395,32 @@ class AttributePropagator {
...
               break;
             }
           }
+        } else if (n->kind() == prim::fork) {
+          applyToSubGraph(n, graph, [this](std::shared_ptr<Graph>& subgraph) {
+            recordReferencedAttrs(subgraph);
+          });
         }
+
+        /*if (n->kind() == prim::CallMethod) {
```
[the remark this answers is not in the document]
Comment (bzinodev) [reply torch/csrc/jit/passes/freeze_module.cpp]: sorry, ghstack

File: torch/csrc/jit/passes/freeze_module.cpp (hunk 2 above)
Comment (bzinodev) [reply torch/csrc/jit/passes/freeze_module.cpp:401]: Fork is not called with  self as param. 
_fork(submodule, foo)

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 8]
```diff
@@ -64,13 +87,17 @@ class AttributePropagator {
...
+      std::string& name,
+      Module& attrModule,
+      std::shared_ptr<Graph>& graph) {
     if (!input->type()->expect<ClassType>()->is_module()) {
       return false;
     }
     Node* node = input->node();
     names_.clear();
-    while (!(node->outputs()[0]->type() == attrModule.type())) {
+    while (!(node->outputs()[0]->type() == graph->inputs()[0]->type())) {
```
[the remark this answers is not in the document]
Comment (bzinodev) [reply torch/csrc/jit/passes/freeze_module.cpp:103]: We use the first argument to  forward(self, ....) or fork(module, ...) to resolve  GetAttr chains

File: torch/csrc/jit/passes/freeze_module.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/freeze_module.cpp:33]: Those regressions should probably be handled separately. I think we should just run DCE - inlining will just adversely affect mobile.

**Review by eellison — COMMENTED** [review]
I think this also needs to add handling for the fork(free_function) case.

Edit: also, seems like this would not be that hard to generalize to the preserve multiple methods / attributes from ehere

File: torch/csrc/jit/passes/freeze_module.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (bzinodev) [reply torch/csrc/jit/passes/freeze_module.cpp:33]: I submitted a separate patch

File: torch/csrc/jit/passes/freeze_module.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (bzinodev) [reply torch/csrc/jit/passes/freeze_module.cpp:339]: Correct

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 9]
```diff
@@ -286,11 +320,38 @@ class AttributePropagator {
               paramConst->debugName());
           n->outputs().at(0)->replaceAllUsesWith(paramConst);
           n->removeAllInputs();
+        } else if (n->kind() == prim::fork) {
+          auto func = [this](std::shared_ptr<Graph>& subgraph) {
```
[the remark this answers is not in the document]
Comment (bzinodev) [reply torch/csrc/jit/passes/freeze_module.cpp:324]: The lambda allows to capture 'this'.  Moreover, These are private methods so the compiler complains :). I think lambda wrap is neat.

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 10]
```diff
@@ -286,11 +320,38 @@ class AttributePropagator {
...
+      std::shared_ptr<Graph>& graph,
+      const std::function<void(std::shared_ptr<Graph>&)>& func) {
+    TORCH_CHECK(n->kind() == prim::fork);
+    auto attrModule = module_;
+    auto node = n->inputs()[0]->node();
+    if (node->kind() != prim::GetAttr)
+      return;
+    auto name = node->s(attr::name);
+    auto input = node->inputs()[0];
+    if (!findConstantAttr(input, name, attrModule, graph)) {
```
Comment (eellison) [inline torch/csrc/jit/passes/freeze_module.cpp:344]: I think even if the module is a constant attribute we still need to run `recordMutableAttrs` on the subgraph here, so that modified attributes/submodules within the subgraph arrer preserved.

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 11]
```diff
@@ -286,11 +320,38 @@ class AttributePropagator {
...
+          applyToSubGraph(n, graph, func);
         }
       }
     }
   }
 
+  void applyToSubGraph(
+      Node* n,
+      std::shared_ptr<Graph>& graph,
+      const std::function<void(std::shared_ptr<Graph>&)>& func) {
```
Comment (eellison) [inline torch/csrc/jit/passes/freeze_module.cpp:336]: It might be easier to follow if this took in an enum of `recordMutableAttrs, recordReferencedAttrs, propagateAttributes` instead of a lambda

**Review by eellison — COMMENTED** [review]
Looks good, I think we still need to make one change to run `recordMutableAttrs` on fork subgraph even if the parent module isn't constant

**Review by eellison — APPROVED** [review]
